### ISSUE TITLE
[CS 1.6] Microphone lags when SetView == true | false

### ISSUE THREAD

**Issue by Limbooc** [ValveSoftware/halflife#2698]
On the video you can see a bug with a microphone in an honest way, when it includes a view from the 3rd person, after switching on, players begin to complain that the microphone is frizzy

Video:
https://www.youtube.com/watch?v=RycOCPgfK54

This error has existed for many years.
used plugin 3rd person view

**/* BIG CREDITS FOR ConnorMcLeod
I Only just modified the front camera and added a scale controlled by a player, so he can zoom in/out! & added a menu*/

```

#include <amxmodx> 
#include <fakemeta> 
#include <hamsandwich> 

#define VERSION "0.0.3" 

#define USE_TOGGLE 3

#define MAX_BACKWARD_UNITS	-200.0
#define MAX_FORWARD_UNITS	200.0

new g_iPlayerCamera[33], Float:g_camera_position[33];

public plugin_init()
{
	register_plugin("Camera View Menu", VERSION, "ConnorMcLeod & Natsheh") 
	
	register_clcmd("say /cam", "camera_menu")
	register_clcmd("say_team /cam", "camera_menu")
	
	register_forward(FM_SetView, "SetView") 
	RegisterHam(Ham_Think, "trigger_camera", "Camera_Think")
}

public camera_menu(id)
{
	if(!is_user_alive(id))
	{
		return 1;
	}
	
	new menu = menu_create("Choose an option!", "cam_m_handler"), sText[48], bool:mode = (g_iPlayerCamera[id] > 0) ? true:false;
	
	formatex(sText, charsmax(sText), "%s \r3RD Person camera!", (mode) ? "\dDisable":"\yEnable")
	menu_additem(menu, sText)
	
	if(mode)
	{
		menu_additem(menu, "Forward Further!")
		menu_additem(menu, "Backward Further!")
	}
	
	menu_display(id, menu)
	return 1;
}

public cam_m_handler(id, menu, item)
{
	if(item == MENU_EXIT)
	{
		menu_destroy(menu)
		return 1;
	}
	
	menu_destroy(menu);
	
	if(g_iPlayerCamera[id] > 0 && item == 0)
	{
		new iEnt = g_iPlayerCamera[id];
		if(pev_valid(iEnt)) engfunc(EngFunc_RemoveEntity, iEnt);
		g_iPlayerCamera[id] = 0;
		g_camera_position[id] = -100.0;
		engfunc(EngFunc_SetView, id, id);
	}
	else
	{
		switch( item )
		{
			case 0:
			{
				g_camera_position[id] = -150.0;
				enable_camera(id)
			}
			case 1: if(g_camera_position[id] < MAX_FORWARD_UNITS) g_camera_position[id] += 50.0;
			case 2: if(g_camera_position[id] > MAX_BACKWARD_UNITS) g_camera_position[id] -= 50.0;
		}
	}
	
	camera_menu(id)
	return 1;
}

public enable_camera(id)
{ 
	if(!is_user_alive(id)) return;
	
	new iEnt = g_iPlayerCamera[id] 
	if(!pev_valid(iEnt))
	{
		static iszTriggerCamera 
		if( !iszTriggerCamera ) 
		{ 
			iszTriggerCamera = engfunc(EngFunc_AllocString, "trigger_camera") 
		} 
		
		iEnt = engfunc(EngFunc_CreateNamedEntity, iszTriggerCamera);
		set_kvd(0, KV_ClassName, "trigger_camera") 
		set_kvd(0, KV_fHandled, 0) 
		set_kvd(0, KV_KeyName, "wait") 
		set_kvd(0, KV_Value, "999999") 
		dllfunc(DLLFunc_KeyValue, iEnt, 0) 
	
		set_pev(iEnt, pev_spawnflags, SF_CAMERA_PLAYER_TARGET|SF_CAMERA_PLAYER_POSITION) 
		set_pev(iEnt, pev_flags, pev(iEnt, pev_flags) | FL_ALWAYSTHINK) 
	
		dllfunc(DLLFunc_Spawn, iEnt)
	
		g_iPlayerCamera[id] = iEnt;
 //   } 	
		new Float:flMaxSpeed, iFlags = pev(id, pev_flags) 
		pev(id, pev_maxspeed, flMaxSpeed)
		
		ExecuteHam(Ham_Use, iEnt, id, id, USE_TOGGLE, 1.0)
		
		set_pev(id, pev_flags, iFlags)
		// depending on mod, you may have to send SetClientMaxspeed here. 
		// engfunc(EngFunc_SetClientMaxspeed, id, flMaxSpeed) 
		set_pev(id, pev_maxspeed, flMaxSpeed)
	}
}

public SetView(id, iEnt) 
{ 
	if(is_user_alive(id))
	{
		new iCamera = g_iPlayerCamera[id] 
		if( iCamera && iEnt != iCamera ) 
		{ 
			new szClassName[16] 
			pev(iEnt, pev_classname, szClassName, charsmax(szClassName)) 
			if(!equal(szClassName, "trigger_camera")) // should let real cams enabled 
			{ 
				engfunc(EngFunc_SetView, id, iCamera) // shouldn't be always needed 
				return FMRES_SUPERCEDE 
			} 
		} 
	} 
	return FMRES_IGNORED 
}

public client_disconnected(id) 
{ 
	new iEnt = g_iPlayerCamera[id];
	if(pev_valid(iEnt)) engfunc(EngFunc_RemoveEntity, iEnt);
	g_iPlayerCamera[id] = 0;
	g_camera_position[id] = -100.0;
} 

public client_putinserver(id) 
{
	g_iPlayerCamera[id] = 0
	g_camera_position[id] = -100.0;
} 

get_cam_owner(iEnt) 
{ 
	new players[32], pnum;
	get_players(players, pnum, "ch");
	
	for(new id, i; i < pnum; i++)
	{ 
		id = players[i];
		
		if(g_iPlayerCamera[id] == iEnt)
		{
			return id;
		}
	}
	
	return 0;
} 

public Camera_Think(iEnt)
{
	static id;
	if(!(id = get_cam_owner(iEnt))) return ;
	
	static Float:fVecPlayerOrigin[3], Float:fVecCameraOrigin[3], Float:fVecAngles[3], Float:fVec[3];
	
	pev(id, pev_origin, fVecPlayerOrigin) 
	pev(id, pev_view_ofs, fVecAngles) 
	fVecPlayerOrigin[2] += fVecAngles[2] 
	
	pev(id, pev_v_angle, fVecAngles) 
	
	angle_vector(fVecAngles, ANGLEVECTOR_FORWARD, fVec);
	static Float:units; units = g_camera_position[id];
	
	//Move back/forward to see ourself
	fVecCameraOrigin[0] = fVecPlayerOrigin[0] + (fVec[0] * units)
	fVecCameraOrigin[1] = fVecPlayerOrigin[1] + (fVec[1] * units) 
	fVecCameraOrigin[2] = fVecPlayerOrigin[2] + (fVec[2] * units) + 15.0
	
	static tr2; tr2 = create_tr2();
	engfunc(EngFunc_TraceLine, fVecPlayerOrigin, fVecCameraOrigin, IGNORE_MONSTERS, id, tr2)
	static Float:flFraction 
	get_tr2(tr2, TR_flFraction, flFraction)
	if( flFraction != 1.0 ) // adjust camera place if close to a wall 
	{
		flFraction *= units;
		fVecCameraOrigin[0] = fVecPlayerOrigin[0] + (fVec[0] * flFraction);
		fVecCameraOrigin[1] = fVecPlayerOrigin[1] + (fVec[1] * flFraction);
		fVecCameraOrigin[2] = fVecPlayerOrigin[2] + (fVec[2] * flFraction);
	}
	
	if(units > 0.0)
	{
		fVecAngles[0] *= fVecAngles[0] > 180.0 ? 1:-1
		fVecAngles[1] += fVecAngles[1] > 180.0 ? -180.0:180.0
	}
	
	set_pev(iEnt, pev_origin, fVecCameraOrigin); 
	set_pev(iEnt, pev_angles, fVecAngles);
	
	free_tr2(tr2);
}

```

**Comment by SamVanheer** [comment]
This happens with `trigger_camera` as well, which also uses `SetView`. There's a problem in the microphone code somewhere that causes this, not sure what it is.

**Comment by afwn90cj93201nixr2e1re** [comment]
At least you should not use nonsteam version with old codecs.
For PoC show video with steam version of cs.

**Comment by Limbooc** [comment]
> At least you should not use nonsteam version with old codecs.
> For PoC show video with steam version of cs.
> 
> Are you sure that this is micro bug? Not trigger camera?



I provide you with a Steam client, I don’t think it’s micrologs or microbags, it happens when I activate trigger_camera. In the end, you can check it yourself using the above plugin

https://www.youtube.com/watch?v=tCr7fdujXwo&feature=youtu.be

**Comment by LucesEnBabel** [comment]
This is an engine error (Not just CS1.6), I remember it happening with Sven Co-op 3.0 / 4.n

**Comment by RauliTop** [comment]
The problem is related when a player using the microphone changes his camera.

Solution: block camera change when the player is speaking. He may change camera when not using microphone.

**Comment by SamVanheer** [comment]
> 
> 
> The problem is related when a player using the microphone changes his camera.
> 
> Solution: block camera change when the player is speaking. He may change camera when not using microphone.

That would break a lot of custom maps and mods. The solution is to fix whatever's causing this in the engine.

**Comment by RauliTop** [comment]
> That would break a lot of custom maps and mods. The solution is to fix whatever's causing this in the engine.

I mean a solution while we are waiting to the fix by Valve. 

Also, you can use the stopsound command after attaching the camera view. That would cause a "cut" if some player is using the microphone at that moment, but you will listen him again without "lag".

I reported it at rehlds project: https://github.com/dreamstalker/rehlds/issues/654

**Comment by Limbooc** [comment]
a temporary solution was:
```
engfunc(EngFunc_SetView, id, id);
client_cmd(id, "stopsound");
```

**Comment by Limbooc** [comment]
> > Это сломало бы много пользовательских карт и модов. Решение состоит в том, чтобы исправить то, что вызывает это в двигателе.
> 
> Я имею в виду решение, пока мы ожидаем исправления от Valve.
> 
> Также вы можете использовать команду Стоп-звук после прикрепления вида с камеры. Это вызвало бы «сокращение», если какой-то игрок использует микрофон в этот момент, но вы будете слушать его снова без «задержки».
> 
> Я сообщил об этом в проекте [rehlds](https://github.com/dreamstalker/rehlds/issues/654) : [dreamstalker / rehlds # 654](https://github.com/dreamstalker/rehlds/issues/654)

client_cmd(id, "stopsound")  Doesn't work, just checked.

**Comment by RauliTop** [comment]
> client_cmd(id, "stopsound") Doesn't work, just checked.

For me it's working.

You have to pay attention to the plugin your using. You must use stopsound when the player use the entity to attach his camera position (that includes when player dead, player spawn again, re-use the entity, stop using the entity...)

I adapted it on a private plugin and it took me some time, including testing.

**Comment by Limbooc** [comment]
> > client_cmd(id, "stopsound") Doesn't work, just checked.
> 
> For me it's working.
> 
> You have to pay attention to the plugin your using. You must use stopsound when the player uses the entity to attach his camera position (that includes when player deads, player spawns again, re-uses the entity, stops using the entity...)
> 
> I adapted it on a private plugin and it took me some time, including testing.


What I did but did not help:
```
if(g_iPlayerCamera[id] > 0 && item == 0)
{
		new iEnt = g_iPlayerCamera[id];
		if(pev_valid(iEnt)) engfunc(EngFunc_RemoveEntity, iEnt);
		g_iPlayerCamera[id] = 0;
		g_camera_position[id] = -100.0;
		client_cmd(id, "stopsound");
		engfunc(EngFunc_SetView, id, id);
		client_cmd(id, "stopsound");
}

public SetView(id, iEnt) 
{ 
	if(is_user_alive(id))
	{
		new iCamera = g_iPlayerCamera[id] 
		if( iCamera && iEnt != iCamera ) 
		{ 
			new szClassName[16] 
			pev(iEnt, pev_classname, szClassName, charsmax(szClassName)) 
			if(!equal(szClassName, "trigger_camera")) // should let real cams enabled 
			{ 
				client_cmd(id, "stopsound");
				engfunc(EngFunc_SetView, id, iCamera) // shouldn't be always needed 
				client_cmd(id, "stopsound");
				return FMRES_SUPERCEDE 
			} 
		} 
	} 
	return FMRES_IGNORED 
}
```


If possible, can you provide your solution

**Comment by RauliTop** [comment]
You have to use stopsound only once AFTER attaching the view.

Also, do it when uses the camera entity:
ExecuteHam(Ham_Use, iEnt, id, id, USE_TOGGLE, 1.0)

**Comment by Limbooc** [comment]
UP!

**Comment by Solokiller** [comment]
This is a duplicate of #1752.

**Comment by kisak-valve** [comment]
Closing in favor of the older issue report.
